draft-ietf-tls-external-psk-importer.md

```diff
@@ -106,18 +123,57 @@ structure as follows:
 
 ~~~
    epskx = HKDF-Extract(0, epsk)
-   ipskx = HKDF-Expand-Label(epskx, "derived psk",
-                             Hash(ImportedIdentity), Hash.length)
+   ipskx = HKDF-Expand-Label(epskx, "external psk importer",
+                             Hash(Context), Hash.length)
+~~~
+
+The `Context` in this case includes the `ImportedIdentity` plus any necessary additional security context information.
+
+~~~
+  struct {
+    opaque channel_binding<1..2^16>;
+    opaque secrets<0..2^16>;
+  } PriorContext
 ~~~
+~~~
+  struct {
+      ImportedIdentity imported_psk;
+      opaque client_id<0..2^16>;
+      PriorContext prior_contexts<0..2^16>;
+  } Context;
+~~~
+
+The `client_id` is an optional field that is required to be unique for each actor that knows the the EPSK.
+This is to prevent Selfie-style reflections.
+This is only necessary in scenarios where more than two actors use the same key, including the case where a single agent will complete PSK handshakes as both the client and the server using the same key.
+The Selfie attack {{!Selfie}} abuses an underlying assumption of TLS, that only one client and one server know a PSK.
+A pair of agents that both act as both client *and* server, and use the same PSK in both roles violate this assumption.
+There are two clients and two servers that know the PSK.
+The agents are thus both vulnerable to reflection attacks where the attacker forces the agent to act as both client and server in a single connection.
+This is because neither agent can distinguish between itself and its peer.
+By hashing a distinguishing value into PSK binder, a server that recieved a reflected `ClientHello` would be unable to verify the binder, and would reject the connection.
+
+For example a distinguishing string for IoT devices could be a MAC address, and for VMs it could be a namespace.
+The decision to use this field, and what it should contain must be agreed OOB.
+N.B. this value is never sent on the wire, as this would identify the client even to passive adverseries.
+
+
+`prior_contexts` is a list of prior security contexts, consisting of channel bindings and any associated keys.
+In the standard case this list will be empty, as the TLS connection is not wrapped in a prior security context.
+However if the OOB PSK was established through a protocol, or series of protocols, that provide a continuing security context, then including the channel binding for these contexts, as well as any keys established, binds the security contexts together.
+This makes it easier to reason formally about the exact properties are provided by the combined series of protocols.
+For example consider chaining together two TLS sessions, but using OOB PSK importers rather than resumption.
+One could then include the prior context `<exporter_key, master_secret>`, where `exporter key` is computed using the exporter interface and a label that uniquely defines this particular setup.
```

Comment:
This feels like it needs to be more precise, especially for this example. Can we add more details here? (What details I am not sure at the moment. I need to chew on this a bit more!)